### PR TITLE
Make it pep 517 and 518 compliant

### PR DESCRIPTION
**Pull request by ocefpaf**
@efiring I also adjusted a few things in the test matrix. If you approve this one the next step the the wheels build infrastructure.

### CODE REVIEW

**Comment by ocefpaf** [comment]
@DocOtak it would be nice to have more people looking into this PR. Let me know if you have any questions.

**Comment by DocOtak** [comment]
Hi @ocefpaf I took a look and here is what I noticed, also I'm not too familiar with pip 517 and 518 (though I've read them).

On a "fresh" 3.8.2 pyenv it was unable to build/install using `python setup.py install`, pip was happy and would build a wheel just fine. I think this is ok, people should be using conda or pip these days.

If they were created by some other process (importing, running pytest), the various __pycache__ dirs and contents would get included in the wheel. My guess is this is undesirable.

I was unable to run `pytest --pyargs gsw` to test the installed version when outside the source directory because the matlab (.m) file was excluded from the wheel. Given that the TEOS-10 website says something like "you must run check functions" for the matlab version, we should probably support the testing of the "installed from wheel" version.

File: .travis.yml [hunk 1]
```diff
@@ -9,22 +9,24 @@ env:
...
   include:
-  - name: "py36-np111"
-    env: PY=3.6 NUMPY=1.11
+  - name: "py36-np116"
+    env: PY=3.6 NUMPY=1.16
   - name: "py36-np117"
-    env: PY=3.6 NUMPY=1.17
-  - name: "py38-np111"
-    env: PY=3.8 NUMPY=1.11
+    env: PY=3.6 NUMPY=1.19
```
Comment (efiring) [inline .travis.yml:15]: The name (numpy version) doesn't match the environment. And, apart from that, I'm also fine with dropping 3.6 entirely at this point.

**Comment by ocefpaf** [comment]
> Hi @ocefpaf I took a look and here is what I noticed, also I'm not too familiar with pip 517 and 518 (though I've read them).
> 
> On a "fresh" 3.8.2 pyenv it was unable to build/install using `python setup.py install`, pip was happy and would build a wheel just fine. I think this is ok, people should be using conda or pip these days.

What was the error? The lack of `numpy` and `cython`? If so, that is OK and we should point people to use `pip` to be PEP 517/518 compliant anyway, but if it was something else I may have made a mistake.

> If they were created by some other process (importing, running pytest), the various `__pycache__` dirs and contents would get included in the wheel. My guess is this is undesirable.

It is undesirable. My setup never created those I'll have to check and prune them from the wheel. Thanks!

> I was unable to run `pytest --pyargs gsw` to test the installed version when outside the source directory because the matlab (.m) file was excluded from the wheel. Given that the TEOS-10 website says something like "you must run check functions" for the matlab version, we should probably support the testing of the "installed from wheel" version.

To be honest I believe that testing is when you are building from source. Those are in the source distribution though. I would like for the wheel to be lighter, no unnecessary npy and no `.m`. With that said, these are pretty light weight now and maybe we should just ship them. I'll push a few more commits later.

**Comment by DocOtak** [comment]
Also I would occasionally get invalid escape sequence deprecation warnings for the two regex strings in check_functions.py https://github.com/TEOS-10/GSW-Python/blob/3635331e9a8f0f45f76f16876a7bd698387494d6/gsw/tests/check_functions.py#L48-L51

Should probably make these `r"raw"` strings...

**Comment by ocefpaf** [comment]
> Also I would occasionally get invalid escape sequence deprecation warnings for the two regex strings in check_functions.py
> 
> https://github.com/TEOS-10/GSW-Python/blob/3635331e9a8f0f45f76f16876a7bd698387494d6/gsw/tests/check_functions.py#L48-L51
> 
> Should probably make these `r"raw"` strings...

Let's do that in a separate PR to avoid mixing the context here. But I agree, they should be raw strings. Fixing those in a moment...

**Comment by DocOtak** [comment]
The error for the setup.py was indeed a lack of numpy.

**Comment by ocefpaf** [comment]
> The error for the setup.py was indeed a lack of numpy.

OK. Then we are "good" finding `numpy` manually in a `setup.py` is deprecated and no longer recommended. `pyproject.toml` does the right thing and the error message is enough to tell people to install `numpy` if they cannot use modern `pip`.

File: .travis.yml (hunk 1 above)
Comment (ocefpaf) [reply .travis.yml:15]: Argh! I checked those a thousand times and missed it anyway. Thanks!

File: setup.cfg [hunk 2]
```diff
@@ -1,3 +1,33 @@
...
+    Programming Language :: Python :: 3.7
+    Programming Language :: Python :: 3.8
+    Topic :: Scientific/Engineering
+
+
+[options]
+zip_safe = False
+install_requires =
+    numpy
+python_requires = >=3.6
```
Comment (efiring) [inline setup.cfg:27]: Is leaving this as '>=3.6' intentional, on the grounds that it should still work fine on 3.6?

File: setup.py [hunk 3]
```diff
@@ -14,12 +13,12 @@
 
 
 # Check Python version.
-if sys.version_info < (3, 5):
+if sys.version_info < (3, 6):
```
Comment (efiring) [inline setup.py:16]: If we are requiring >= 3.6 or >= 3.7, do we still need this check?

Comment (ocefpaf) [reply setup.py:16]: Nope. Let me remove that.

Comment (ocefpaf) [reply setup.py]: I removed the pip check too and let `pyproject.toml` handle the setup dependencies.

**Comment by efiring** [comment]
Sorry, I got distracted.  Let's give it a try, and iterate if needed.

**Comment by ocefpaf** [comment]
> Sorry, I got distracted. Let's give it a try, and iterate if needed.

No problem. I won't be able to continue this in the next weeks due to an event I'm participating. I'm planning on getting back to it as soon as possible though.

**Pull request merged**